Day8

wontheone1 opened the pull request:
Day4 PR이 머지되면 리베이스 할 예정입니다.

https://github.com/wontheone1/adventOfCode_Rescript/pull/2

Day8 긴호흡으로 한번에 풀고 미뤘던 테스트는 마지막 커밋에 올렸습니다.

wontheone1 committed:
Start Day8, split instructions by newlines

wontheone1 committed:
Define OptionResult module for converting between option and result

wontheone1 committed:
Add utility functions

wontheone1 committed:
Create instruction tuples from instruction lines

wontheone1 committed:
Use instructionKind type

wontheone1 committed:
Remove unnecessary helper function for parseInstructions

wontheone1 committed:
Solve Day8

wontheone1 committed:
Move Day4 files to Day4 folder

wontheone1 committed:
Move Day8 file to Day8 folder

wontheone1 committed:
Split Day8 solution files to multiple pieces

wontheone1 committed:
Define initialState from makeInitialState function

wontheone1 committed:
Test Day8 solution

kimchhickey commented on src/Day8/InstructionRunner.res:
```diff
@@ -0,0 +1,79 @@
+open Belt
+open InstructionModel
+
+let rec runInstruction = (appState: InstructionModel.appState, instructions) => {
+  if appState.isFinished {
+    appState
+  } else {
+    let instructionToRun = instructions[appState.currentInstructionIndex]
+    switch instructionToRun {
+    | None => {
+        ...appState,
+        isFinished: true,
+      }
+    | Some(instruction) => {
+        let numberOfCurrentInstructionVisited = Option.getWithDefault(
+          appState.numberOfInstructionsVisited[appState.currentInstructionIndex],
+          0,
+        )
+        if numberOfCurrentInstructionVisited > 0 {
```
이 부분을 상태 머신과 재귀 함수로 분리해보면 어떨까요?
만약에 클로져 코드로 짰으면 아래처럼 했을 것 같거든요.
pred를 받는 take-while과, state를 조작하는 함수로 나누는 것이죠!!
```clojure
(take-while terminate-fn (iterate ->next-state initial-state))
```

wontheone1 replied:
아하 이 클로져 원 라이너 너무 좋네요 ㅎㅎㅎ

한번 분리해볼게요

wontheone1 replied:
마지막 커밋으로 분리했습니다. `instructions`를 `appState` 안에 넣어볼까 고민하다가, instructions는 한 러닝 세션에서 업데이트 되는게 아니라 그냥 따로 뒀습니다. 어떻게 해도 괜찮을 것 같아요.

kimchhickey commented on src/Day8/InstructionParser.res:
```diff
@@ -0,0 +1,34 @@
+  | "nop" => NoOp
+  | "acc" => Accumulate(arg)
+  | "jmp" => Jump(arg)
+  | _ => raise(InvalidInstruction)
```
너무 좋은데요 👍

wontheone1 replied:
감사합니다 :)

kimchhickey commented on src/Day8/InstructionParser.res:
```diff
@@ -0,0 +1,34 @@
+  }
+}
+
+let parseInstructions = (instructionLines: Js.String2.t): array<
```
string과 Js.String2.t간에는 상호호환성이 있어서 string으로 쓰는 편입니다.
```rescript
string => array<InstructionModel.instructionKind>
```

wontheone1 replied:
아하, 고쳤어요

kimchhickey commented on src/Day8/InstructionParser.res:
```diff
@@ -0,0 +1,34 @@
+  ->ArrayUtil.removeNones
+  ->Array.map((instructionLine: Js.String2.t) => {
+    let instruction = Js.String2.splitByRe(instructionLine, %re("/\s/"))->ArrayUtil.removeNones
+    if Array.length(instruction) != 2 {
+      raise(InvalidInstruction)
+    }
+    let maybeOpArg = Option.flatMap(instruction[0], op =>
+      Option.map(instruction[1], arg => instructionFromInstructionTuple((op, arg)))
+    )
+    switch maybeOpArg {
+    | None => raise(InvalidInstruction)
+    | Some(instruction) => instruction
+    }
```
요렇게 줄여볼 수도 있을 것 같아요.
```rescript
switch (instruction[0], instruction[1]) {
| (Some(op), Some(arg)) => instructionFromInstructionTuple((op, arg))
| _ => raise(InvalidInstruction)
}
```

wontheone1 replied:
아하, 이 생각을 못했을까요ㅎㅎ 좋은데요, 고쳤습니다.
`Array.length(instruction) != 2` 체크는 그냥 놔뒀습니다. 왜냐하면 instruction[0], instruction[1] 이거 두개가 올바른 인풋이라도 그 라인에 뭔가 불필요한 게 더 있다면 인풋의 질이 나쁘다고 볼 수 있기 때문입니다. e.g,
```
nop +0 acd
acc +1 1111
jmp +4 minus
```

wontheone1 replied:
이경우 에러 타입을 따로 하나 정의해줘도 좋을 것 같네요.

wontheone1 committed:
Use string type instead of Js.String2.t

wontheone1 committed:
Switch on op, arg tuple instead of flatmap mapping with Belt.Option

wontheone1 committed:
Inline accessing current instruction

wontheone1 committed:
Refactor reducer of runInstruction

wontheone1 committed:
Terminate runInstruction function based on terminationTest function

and remove isFinished field from appState since terminationTest function determines it

wontheone1 commented on src/Day8/InstructionRunner.res:
```diff
@@ -0,0 +1,78 @@
+
+      switch instruction {
+      | NoOp => {
+          ignore(
+            appState.numberOfInstructionsVisited[
+              appState.currentInstructionIndex
+            ] =
+              numberOfCurrentInstructionVisited + 1,
+          )
```
여기 ignore에 넣은 부분을 부작용 없이 맵으로 할수도 있는데 뮤터블 어레이라 맵을하는 거보다 이렇게 변경하는 것이 성능적 특질이 좋을 것 같긴해서 순수 함수로 굳이 고치지 않았습니다. 어떻게 하는 걸 선호하실 지 모르겠네요. @kimchhickey

kimchhickey replied:
오... 저도 이런 접근에 동의합니다.
함수 전체의 순수성만 보장이 되면 local scope에서는 효율을 위해서 mutable을 사용할 수 있다고 생각해요. (마치 클로져의 transient 같이!)

이런 것을 허용한다고 하면 visited를 추가하는 부분은 모든 instruction에서 공통으로 사용되니까, 위쪽으로 추출할 수 있지 않을까요?

kimchhickey replied:
그리고 사소한거지만 ignore에서는 pipe 연산자를 쓰면 내용을 먼저 읽을 수 있어서 가독성이 좀 올라갑니다.
```rescript
ignore(myCode)

vs

myCode->ignore
```

kimchhickey commented on src/Day8/InstructionRunner.res:
```diff
@@ -0,0 +1,78 @@
+open Belt
+open InstructionModel
+
+let reducer = (action, appState) => {
```
오 reducer라는 이름 참 좋습니다!
```rescript
(action, state) => state
```

wontheone1 replied:
아 넵! @kimchhickey

https://www.youtube.com/watch?v=Hv_PhrfwerQ 여기 상태 머신에 관한 테크탁을 보다가 이렇게 하는 걸 보고 따라해봤어요

kimchhickey reviewed (commented):
잘하셔서 별로 리뷰 드릴게 없네요 👍 
고생 많이하셨습니다!

wontheone1 commented:
> ignore에서는 pipe 연산자를 쓰면 내용을 먼저 읽을 수 있어서 가독성이 좀 올라갑니다.

@kimchhickey 
사실 쓰려고 했는데 어싸인먼트 자체를 ignore 에 보내는데 약간 문제가 있었어요
![Screenshot 2021-03-01 at 15 01 20](https://user-images.githubusercontent.com/11784756/109500681-373c0080-7a9f-11eb-8a72-0f4d0475d494.png)

조금 고민하다가 {} 이용해서 어싸인먼트를 한덩이로 먼저 묶고 이그노어에 보내는 방법을 알아냈어요! 고민하니 다 수가 있네요 감사합니다.

wontheone1 committed:
Use -> to ignore for readability

wontheone1 commented:
> 위쪽으로 추출할 수 있지 않을까요?

위쪽에 로컬 도움 함수로 따로 뽑아냈습니다!
